ttnn/cpp/ttnn/operations/normalization/softmax/softmax_compiler_interface.hpp

```diff
@@ -0,0 +1,30 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <exception>
+#include <functional>
+
+#include "third_party/json/json.hpp"
+#include "tt_metal/common/logger.hpp"
+#include "ttnn/compiler_interface/compiler_interface.hpp"
+#include "ttnn/graph/graph_processor.hpp"
+#include "ttnn/graph/graph_trace_utils.hpp"
+#include "ttnn/operations/normalization/softmax/softmax.hpp"
+
+namespace ttnn::compiler_interface {
+
+QueryResponse softmax_op_constraints(
```

Comment:
I don't think it's a good idea to manually write an implementation of this for each op. I believe you should be able to make a single variadic template function, which would do this. Basically you only need to replace each TensorSpec argument with a result of `create_device_tensor`, and then just decorate the call with `begin_graph_capture`/`end_graph_capture`.

Reply:
thanks, I'll give it a try.

Reply:
It seems like this function should work for any op:
```cpp
template<typename Op, typename... Args>
auto query_constraints(Op op, Device* device, Args&&... args) {
    auto update_arg = [device](auto&& arg) {
        if constexpr (std::is_same_v<std::decay_t<decltype(arg)>, TensorSpec>) {
            return create_device_tensor(arg, device);
        } else {
            return std::forward<decltype(arg)>(arg);
        }
    };
    auto transformed_args = std::make_tuple(transform_arg(std::forward<Args>(args))...);
    ttnn::graph::GraphProcessor::begin_graph_capture(ttnn::graph::GraphProcessor::RunMode::NO_DISPATCH);
    std::apply(op, transformed_args);
    return ttnn::graph::GraphProcessor::end_graph_capture();
}
```